router/xgress/payload_buffer.go

```diff
@@ -49,15 +64,30 @@ func (controller *PayloadBufferController) BufferForSession(sessionId *identity.
 		return i.(*PayloadBuffer)
 	}
 
-	buffer := NewPayloadBuffer(sessionId, controller.forwarder)
+	buffer := NewPayloadBuffer(controller, sessionId, controller.forwarder)
 	controller.buffers.Set(bufferId, buffer)
+
```

Comment:
Maybe use cmap.Upsert here? There's a chance of a read/write conflict here. Not sure how likely it actually is, given it would have to be two different things with the same session, and sequencing would likely prevent that happening